File: go/roothash/api/commitment/pool.go
package commitment

import (
	"context"
	"math"

	beacon "github.com/oasisprotocol/oasis-core/go/beacon/api"
	"github.com/oasisprotocol/oasis-core/go/common/crypto/hash"
	"github.com/oasisprotocol/oasis-core/go/common/crypto/signature"
	"github.com/oasisprotocol/oasis-core/go/common/errors"
	"github.com/oasisprotocol/oasis-core/go/common/logging"
	"github.com/oasisprotocol/oasis-core/go/common/node"
	p2pError "github.com/oasisprotocol/oasis-core/go/p2p/error"
	registry "github.com/oasisprotocol/oasis-core/go/registry/api"
	"github.com/oasisprotocol/oasis-core/go/roothash/api/block"
	"github.com/oasisprotocol/oasis-core/go/roothash/api/message"
	scheduler "github.com/oasisprotocol/oasis-core/go/scheduler/api"
)

// moduleName is the module name used for namespacing errors.
const moduleName = "roothash/commitment"

// nolint: revive
var (
	ErrNoRuntime              = errors.New(moduleName, 1, "roothash/commitment: no runtime configured")
	ErrNoCommittee            = errors.New(moduleName, 2, "roothash/commitment: no committee configured")
	ErrInvalidCommitteeKind   = errors.New(moduleName, 3, "roothash/commitment: invalid committee kind")
	ErrRakSigInvalid          = errors.New(moduleName, 4, "roothash/commitment: batch RAK signature invalid")
	ErrNotInCommittee         = errors.New(moduleName, 5, "roothash/commitment: node not part of committee")
	ErrAlreadyCommitted       = errors.New(moduleName, 6, "roothash/commitment: node already sent commitment")
	ErrNotBasedOnCorrectBlock = errors.New(moduleName, 7, "roothash/commitment: submitted commitment is not based on correct block")
	ErrDiscrepancyDetected    = errors.New(moduleName, 8, "roothash/commitment: discrepancy detected")
	ErrStillWaiting           = errors.New(moduleName, 9, "roothash/commitment: still waiting for commits")
	ErrInsufficientVotes      = errors.New(moduleName, 10, "roothash/commitment: insufficient votes to finalize discrepancy resolution round")
	ErrBadExecutorCommitment  = errors.New(moduleName, 11, "roothash/commitment: bad executor commitment")
	// Error code 12 is reserved for future use.
	ErrInvalidMessages = p2pError.Permanent(errors.New(moduleName, 13, "roothash/commitment: invalid messages"))
	// Error code 14 is reserved for future use.
	ErrTimeoutNotCorrectRound = errors.New(moduleName, 15, "roothash/commitment: timeout not for correct round")
	ErrInvalidRound           = errors.New(moduleName, 16, "roothash/commitment: invalid round")
	ErrNoSchedulerCommitment  = errors.New(moduleName, 17, "roothash/commitment: no scheduler commitment")
	ErrBadSchedulerCommitment = errors.New(moduleName, 18, "roothash/commitment: bad scheduler commitment")
)

// LogEventDiscrepancyMajorityFailure is a log event value that dependency resolution with majority failure.
const LogEventDiscrepancyMajorityFailure = "pool/discrepancy_majority_failure"

var logger *logging.Logger = logging.GetLogger("roothash/commitment/pool")

// NodeLookup is an interface for looking up registry node descriptors.
type NodeLookup interface {
	// Node looks up a node descriptor.
	Node(ctx context.Context, id signature.PublicKey) (*node.Node, error)
}

// MessageValidator is an arbitrary function that validates messages for validity. It can be used
// for gas accounting.
type MessageValidator func(msgs []message.Message) error

// Pool is a serializable pool of scheduler commitments that can be used to perform discrepancy
// detection and resolution.
//
// The pool is not safe for concurrent use.
type Pool struct {
	// HighestRank is the rank of the highest-ranked scheduler among those who have submitted
	// a commitment for their own proposal. The maximum value indicates that no scheduler
	// has submitted a commitment.
	HighestRank uint64 `json:"highest_rank,omitempty"`
	// SchedulerCommitments is a map that groups scheduler commitments and worker votes
	// by the scheduler's rank.
	SchedulerCommitments map[uint64]*SchedulerCommitment `json:"scheduler_commitments,omitempty"`
	// Discrepancy is a flag signalling that a discrepancy has been detected.
	Discrepancy bool `json:"discrepancy,omitempty"`
}

// NewPool creates a new pool without any commitments and with .
func NewPool() *Pool {
	return &Pool{
		HighestRank: math.MaxUint64,
	}
}

// VerifyExecutorCommitment verifies the given executor commitment.
func VerifyExecutorCommitment( // nolint: gocyclo
	ctx context.Context,
	blk *block.Block,
	rt *registry.Runtime,
	epoch beacon.EpochTime,
	commit *ExecutorCommitment,
	msgValidator MessageValidator,
	nl NodeLookup,
) error {
	// Check executor commitment signature.
	if err := commit.Verify(rt.ID); err != nil {
		return p2pError.Permanent(err)
	}

	// Validate executor commitment.
	if err := commit.ValidateBasic(); err != nil {
		logger.Debug("executor commitment validate basic error",
			"err", err,
		)
		return ErrBadExecutorCommitment
	}

	// Check if the block is based on the previous block.
	if !commit.Header.Header.IsParentOf(&blk.Header) {
		logger.Debug("executor commitment is not based on correct block",
			"node_id", commit.NodeID,
			"expected_previous_hash", blk.Header.EncodedHash(),
			"previous_hash", commit.Header.Header.PreviousHash,
		)
		return ErrNotBasedOnCorrectBlock
	}

	// TODO: Check for evidence of equivocation (oasis-core#3685).

	switch commit.IsIndicatingFailure() {
	case true:
		// Prevent schedulers to submit failures.
		if commit.NodeID.Equal(commit.Header.SchedulerID) {
			logger.Debug("executor commitment scheduler is not allowed to submit a failure",
				"node_id", commit.NodeID,
			)
			return ErrBadExecutorCommitment
		}
	case false:
		// Verify RAK-attestation.
		if rt.TEEHardware != node.TEEHardwareInvalid {
			n, err := nl.Node(ctx, commit.NodeID)
			if err != nil {
				// This should never happen as nodes cannot disappear mid-epoch.
				logger.Warn("unable to fetch node descriptor to verify RAK-attestation",
					"err", err,
					"node_id", commit.NodeID,
				)
				return ErrNotInCommittee
			}

			ad := rt.ActiveDeployment(epoch)
			if ad == nil {
				// This should never happen as we prevent this elsewhere.
				logger.Error("no active deployment",
					"runtime_id", rt.ID,
					"node_id", commit.NodeID,
					"deployments", rt.Deployments,
				)
				return ErrNoRuntime
			}

			nodeRt := n.GetRuntime(rt.ID, ad.Version)
			if nodeRt == nil {
				// We currently prevent this case throughout the rest of the system.
				// Still, it's prudent to check.
				logger.Warn("committee member not registered with this runtime",
					"runtime_id", rt.ID,
					"node_id", commit.NodeID,
				)
				return ErrNotInCommittee
			}

			if nodeRt.Capabilities.TEE == nil {
				// This should never happen as we prevent this elsewhere.
				logger.Error("node doesn't have TEE capability",
					"runtime_id", rt.ID,
					"node_id", commit.NodeID,
				)
				return ErrRakSigInvalid
			}

			if err = commit.Header.VerifyRAK(nodeRt.Capabilities.TEE.RAK); err != nil {
				return ErrRakSigInvalid
			}
		}

		// Check emitted runtime messages.
		switch commit.NodeID.Equal(commit.Header.SchedulerID) {
		case true:
			// The transaction scheduler can include messages.
			if uint32(len(commit.Messages)) > rt.Executor.MaxMessages {
				logger.Debug("executor commitment from scheduler has too many messages",
					"node_id", commit.NodeID,
					"num_messages", len(commit.Messages),
					"max_messages", rt.Executor.MaxMessages,
				)
				return ErrInvalidMessages
			}
			if h := message.MessagesHash(commit.Messages); !h.Equal(commit.Header.Header.MessagesHash) {
				logger.Debug("executor commitment from scheduler has invalid messages hash",
					"node_id", commit.NodeID,
					"expected_hash", h,
					"messages_hash", commit.Header.Header.MessagesHash,
				)
				return ErrInvalidMessages
			}

			// Perform custom message validation and propagate the error unchanged.
			if msgValidator != nil && len(commit.Messages) > 0 {
				err := msgValidator(commit.Messages)
				if err != nil {
					logger.Debug("executor commitment from scheduler has invalid messages",
						"err", err,
						"node_id", commit.NodeID,
					)
					return err
				}
			}
		case false:
			// Other workers cannot include any messages.
			if len(commit.Messages) > 0 {
				logger.Debug("executor commitment from non-scheduler contains messages",
					"node_id", commit.NodeID,
					"num_messages", len(commit.Messages),
				)
				return ErrInvalidMessages
			}
		}
	}

	return nil
}

// AddVerifiedExecutorCommitment adds a verified executor commitment to the pool.
func (p *Pool) AddVerifiedExecutorCommitment(c *scheduler.Committee, ec *ExecutorCommitment) error {
	// Enforce specific roles based on current discrepancy state.
	switch {
	case !p.Discrepancy && !c.IsMember(ec.NodeID):
		// Discrepancy detection accepts commitments arriving in any order, e.g., a backup worker
		// can submit a commitment even before there is a discrepancy.
		logger.Debug("node is not in the committee",
			"round", ec.Header.Header.Round,
			"node_id", ec.NodeID,
		)
		return ErrNotInCommittee
	case p.Discrepancy && !c.IsBackupWorker(ec.NodeID):
		// Discrepancy resolution accepts commitments only from backup workers to prevent workers
		// from improving their liveness statistics.
		logger.Debug("node is not a backup worker",
			"round", ec.Header.Header.Round,
			"node_id", ec.NodeID,
		)
		return ErrBadExecutorCommitment
	}

	// Ensure that the scheduler is allowed to schedule transactions.
	rank, ok := c.SchedulerRank(ec.Header.Header.Round, ec.Header.SchedulerID)
	if !ok {
		// Reject commitments with invalid schedulers.
		logger.Debug("executor commitment's scheduler is not in the committee",
			"round", ec.Header.Header.Round,
			"node_id", ec.NodeID,
			"scheduler_id", ec.Header.SchedulerID,
			"rank", rank,
		)
		return ErrBadExecutorCommitment
	}

	// Prioritize commitments.
	switch {
	case rank > p.HighestRank:
		// Reject commitments with higher ranking.
		logger.Debug("executor commitment's scheduler has worse ranking",
			"round", ec.Header.Header.Round,
			"node_id", ec.NodeID,
			"scheduler_id", ec.Header.SchedulerID,
			"rank", rank,
			"highest_rank", p.HighestRank,
		)
		return ErrBadExecutorCommitment
	case rank != p.HighestRank && p.Discrepancy:
		// Prevent placing commitments with different rank during discrepancy resolution.
		logger.Debug("executor commitment's scheduler rank does not match",
			"round", ec.Header.Header.Round,
			"node_id", ec.NodeID,
			"scheduler_id", ec.Header.SchedulerID,
			"rank", rank,
			"highest_rank", p.HighestRank,
		)
		return ErrBadExecutorCommitment
	case rank < p.HighestRank:
		// Update the pool when a scheduler with a superior rank commits.
		if !ec.NodeID.Equal(ec.Header.SchedulerID) {
			break
		}

		p.HighestRank = rank

		// Drop commitments with higher ranking.
		for r := range p.SchedulerCommitments {
			if r > p.HighestRank {
				delete(p.SchedulerCommitments, r)
			}
		}
	}

	// Add commitment if the node hasn't submitted one.
	if p.SchedulerCommitments == nil {
		p.SchedulerCommitments = make(map[uint64]*SchedulerCommitment)
	}
	sc, ok := p.SchedulerCommitments[rank]
	if !ok {
		sc = &SchedulerCommitment{}
		p.SchedulerCommitments[rank] = sc
	}

	return sc.Add(ec)
}

// ProcessCommitments performs discrepancy detection or resolution.
func (p *Pool) ProcessCommitments(c *scheduler.Committee, allowedStragglers uint16, timeout bool) (*SchedulerCommitment, error) {
	sc, err := p.processCommitments(c, allowedStragglers, timeout)
	switch err {
	case ErrDiscrepancyDetected:
		// Switch to discrepancy resolution.
		p.Discrepancy = true

		// Drop commitments with different ranking.
		for r := range p.SchedulerCommitments {
			if r != p.HighestRank {
				delete(p.SchedulerCommitments, r)
			}
		}
	}

	return sc, err
}

func (p *Pool) processCommitments(c *scheduler.Committee, allowedStragglers uint16, timeout bool) (*SchedulerCommitment, error) { // nolint: gocyclo
	// Ensure we have at least scheduler's vote.
	sc, ok := p.SchedulerCommitments[p.HighestRank]
	switch {
	case !ok && timeout:
		// The round timer expired, but the schedulers haven't submitted any commitments.
		return nil, ErrNoSchedulerCommitment
	case !ok:
		// Wait for additional commitments or until the round timer expires.
		return nil, ErrStillWaiting

	}

	// Gather votes.
	var total, commits, failures int
	votes := make(map[hash.Hash]int)

	for _, n := range c.Members {
		switch {
		case !p.Discrepancy && n.Role != scheduler.RoleWorker:
			continue
		case p.Discrepancy && n.Role != scheduler.RoleBackupWorker:
			continue
		}

		total++
		vote, ok := sc.Votes[n.PublicKey]
		switch {
		case !ok:
			continue
		case vote == nil:
			failures++
		default:
			votes[*vote]++
		}
		commits++

		// Early discrepancy detection.
		switch {
		case p.Discrepancy:
			// Discrepancy resolution already started.
			continue
		case len(votes) <= 1 && failures <= int(allowedStragglers):
			// Discrepancy not detected.
			continue
		case p.HighestRank > 0 && !timeout:
			// To ensure fairness, the backup schedulers are always required to wait for a round
			// timeout. Without this delay, there is a risk that they could bypass the scheduling
			// order by convincing a few other nodes to instantly publish failures or discrepant
			// commitments.
			//
			// In the former scenario, the colliding nodes might avoid punishment because nodes
			// posting a failure are merely marked as 'not alive' for the current round, and no
			// slashing occurs. In the latter case, the colliding nodes may be penalized, but
			// the potential gains from being a scheduler could outweigh the losses.
			continue
		}

		// The committee is neither unanimous nor is the number of valid commitments expected to
		// surpass the required threshold due to numerous failures. Start discrepancy resolution.
		return nil, ErrDiscrepancyDetected
	}

	switch p.Discrepancy {
	case false:
		// Discrepancy detection.
		if len(votes) > 1 || failures > int(allowedStragglers) {
			// Early discrepancy detection will always identify discrepancies for the primary
			// scheduler. However, for the backup schedulers, this will only be the case
			// during timeouts. In other cases, the backup schedulers must await a round timeout.
			return nil, ErrStillWaiting
		}

		// Check if the majority has been reached.
		required := total - int(allowedStragglers)
		for _, v := range votes {
			// The map should have exactly one key/value pair, which indicates how many votes
			// the scheduler's commitment has received.
			required -= v
		}

		switch {
		case required > 0 && timeout:
			// The round timer expired, but the required number of votes hasn't been reached.
			return nil, ErrDiscrepancyDetected
		case required > 0:
			// Wait for additional commitments or until the round timer expires.
			return nil, ErrStillWaiting
		}
	case true:
		// Discrepancy resolution.
		required := total/2 + 1
		remaining := total - commits

		// Find the commit with the highest number of votes.
		var (
			hash hash.Hash
			best int
		)
		for h, v := range votes {
			if v > best {
				hash = h
				best = v
			}
		}

		switch {
		case best+remaining < required:
			// Fail the round if the majority cannot be reached due to insufficient votes remaining
			// (e.g. too many nodes have failed).
			return nil, ErrInsufficientVotes
		case best < required && timeout:
			// The round timer expired, but the majority hasn't been reached.
			return nil, ErrInsufficientVotes
		case best < required:
			// Wait for additional commitments or until the round timer expires.
			return nil, ErrStillWaiting
		case hash != sc.Commitment.ToVote():
			// The scheduler's commitment hasn't received the majority of the votes.
			return nil, ErrBadSchedulerCommitment
		}
	}

	return sc, nil
}

File: runtime/src/consensus/roothash/commitment/pool.rs
use std::collections::{HashMap, HashSet};

use anyhow::Result;

use crate::{
    common::crypto::{hash::Hash, signature::PublicKey},
    consensus::{
        registry::{Node, Runtime, TEEHardware},
        roothash::{Block, Error, Message, OpenCommitment},
        scheduler::{Committee, CommitteeKind, Role},
    },
};

use super::ExecutorCommitment;

/// A trait for looking up registry node descriptors.
pub trait NodeLookup {
    fn node(&self, id: PublicKey) -> Result<Node, Error>;
}

/// A trait that validates messages for validity. It can be used for gas accounting.
pub trait MessageValidator {
    fn validate(&self, msgs: &[Message]) -> Result<()>;
}

impl<F> MessageValidator for F
where
    F: Fn(&[Message]) -> Result<()>,
{
    fn validate(&self, msgs: &[Message]) -> Result<()> {
        (*self)(msgs)
    }
}

/// A pool of commitments that can be used to perform
/// discrepancy detection.
///
/// The pool is not safe for concurrent use.
pub struct Pool {
    /// The runtime descriptor this pool is collecting
    /// the commitments for.
    runtime: Runtime,
    /// The committee this pool is collecting the commitments for.
    committee: Committee,
    /// The current protocol round.
    round: u64,
    // The commitments in the pool iff Committee.Kind
    // is scheduler.KindComputeExecutor.
    execute_commitments: HashMap<PublicKey, ExecutorCommitment>,
    // A flag signalling that a discrepancy has been detected.
    discrepancy: bool,
    // The time when the next call to TryFinalize(true) should
    // be scheduled to be executed. Zero means that no timeout is to be scheduled.
    _next_timeout: i64,

    // A cached committee member set. It will be automatically
    // constructed based on the passed Committee.
    member_set: HashSet<PublicKey>,
    // A cached committee worker set. It will be automatically
    // constructed based on the passed Committee.
    _worker_set: HashSet<PublicKey>,
}

impl Pool {
    /// Creates a new pool.
    pub fn new(runtime: Runtime, committee: Committee, round: u64) -> Self {
        let mut member_set = HashSet::new();
        let mut _worker_set = HashSet::new();

        for m in &committee.members {
            member_set.insert(m.public_key);
            if m.role == Role::Worker {
                _worker_set.insert(m.public_key);
            }
        }

        Pool {
            runtime,
            committee,
            round,
            execute_commitments: HashMap::new(),
            discrepancy: false,
            _next_timeout: 0,
            member_set,
            _worker_set,
        }
    }

    fn is_member(&self, id: &PublicKey) -> bool {
        self.member_set.contains(id)
    }

    fn _is_worker(&self, id: &PublicKey) -> bool {
        self._worker_set.contains(id)
    }

    fn is_scheduler(&self, id: &PublicKey) -> bool {
        if let Ok(scheduler) = self.committee.transaction_scheduler(self.round) {
            return &scheduler.public_key == id;
        }
        false
    }

    /// Verifies and adds a new executor commitment to the pool.
    fn add_verified_executor_commitment(
        &mut self,
        blk: &Block,
        nl: &impl NodeLookup,
        msg_validator: &impl MessageValidator,
        commit: ExecutorCommitment,
    ) -> Result<()> {
        if self.committee.kind != CommitteeKind::ComputeExecutor {
            return Err(Error::InvalidCommitteeKind.into());
        }

        // Ensure that the node is actually a committee member. We do not enforce specific
        // roles based on current discrepancy state to allow commitments arriving in any
        // order (e.g., a backup worker can submit a commitment even before there is a
        // discrepancy).
        if !self.is_member(&commit.node_id) {
            return Err(Error::NotInCommittee.into());
        }

        // Ensure the node did not already submit a commitment.
        if self.execute_commitments.contains_key(&commit.node_id) {
            return Err(Error::AlreadyCommitted.into());
        }

        if self.round != blk.header.round {
            return Err(Error::InvalidRound.into());
        }

        // Check if the block is based on the previous block.
        if !commit.header.header.is_parent_of(&blk.header) {
            return Err(Error::NotBasedOnCorrectBlock.into());
        }

        if commit.validate_basic().is_err() {
            return Err(Error::BadExecutorCommitment.into());
        }

        // TODO: Check for evidence of equivocation (oasis-core#3685).

        if !commit.is_indicating_failure() {
            // Verify RAK-attestation.
            if self.runtime.tee_hardware != TEEHardware::TEEHardwareInvalid {
                let n = nl.node(commit.node_id).map_err(|_|
                    // This should never happen as nodes cannot disappear mid-epoch.
                    Error::NotInCommittee)?;

                let ad = self
                    .runtime
                    .active_deployment(self.committee.valid_for)
                    .ok_or(
                        // This should never happen as we prevent this elsewhere.
                        Error::NoRuntime,
                    )?;

                let rt = n.get_runtime(&self.runtime.id, &ad.version).ok_or(
                    // We currently prevent this case throughout the rest of the system.
                    // Still, it's prudent to check.
                    Error::NotInCommittee,
                )?;

                let tee = rt.capabilities.tee.ok_or(
                    // This should never happen as we prevent this elsewhere.
                    Error::RakSigInvalid,
                )?;

                commit
                    .header
                    .verify_rak(tee.rak)
                    .map_err(|_| Error::RakSigInvalid)?;
            }

            // Check emitted runtime messages.
            match self.is_scheduler(&commit.node_id) {
                true => {
                    // The transaction scheduler can include messages.
                    if commit.messages.len() as u32 > self.runtime.executor.max_messages {
                        return Err(Error::InvalidMessages.into());
                    }

                    let messages_hash = commit
                        .header
                        .header
                        .messages_hash
                        .ok_or(Error::InvalidMessages)?;
                    let h = Message::messages_hash(&commit.messages);
                    if h != messages_hash {
                        return Err(Error::InvalidMessages.into());
                    }

                    // Perform custom message validation and propagate the error unchanged.
                    if !commit.messages.is_empty() {
                        msg_validator.validate(&commit.messages)?;
                    }
                }
                false => {
                    // Other workers cannot include any messages.
                    if !commit.messages.is_empty() {
                        return Err(Error::InvalidMessages.into());
                    }
                }
            }
        }

        self.execute_commitments.insert(commit.node_id, commit);

        Ok(())
    }

    /// Verifies and adds a new executor commitment to the pool.
    pub fn add_executor_commitment(
        &mut self,
        blk: &Block,
        nl: &impl NodeLookup,
        commit: ExecutorCommitment,
        msg_validator: &impl MessageValidator,
        chain_context: &String,
    ) -> Result<()> {
        // Check executor commitment signature.
        commit.verify(&self.runtime.id, chain_context)?;

        self.add_verified_executor_commitment(blk, nl, msg_validator, commit)
    }

    /// Performs a single round of commitment checks. If there are enough commitments
    /// in the pool, it performs discrepancy detection or resolution.
    pub fn process_commitments(&mut self, did_timeout: bool) -> Result<&dyn OpenCommitment> {
        if self.committee.kind != CommitteeKind::ComputeExecutor {
            panic!(
                "roothash/commitment: unknown committee kind: {:?}",
                self.committee.kind
            );
        }

        #[derive(Default)]
        struct Vote<'a> {
            commit: Option<&'a ExecutorCommitment>,
            tally: u16,
        }

        let mut total = 0;
        let mut commits = 0;
        let mut failures = 0;

        // Gather votes.
        let mut votes: HashMap<Hash, Vote> = HashMap::new();
        for n in &self.committee.members {
            if !self.discrepancy && n.role != Role::Worker {
                continue;
            }
            if self.discrepancy && n.role != Role::BackupWorker {
                continue;
            }

            total += 1;
            let commit = match self.execute_commitments.get(&n.public_key) {
                Some(commit) => commit,
                None => continue,
            };
            commits += 1;

            if commit.is_indicating_failure() {
                failures += 1;
                continue;
            }

            let k = commit.to_vote();
            match votes.get_mut(&k) {
                Some(v) => v.tally += 1,
                None => {
                    votes.insert(
                        k,
                        Vote {
                            tally: 1,
                            commit: Some(commit),
                        },
                    );
                }
            }

            if !self.discrepancy && votes.len() > 1 {
                self.discrepancy = true;
                return Err(Error::DiscrepancyDetected.into());
            }
        }

        // Determine whether the proposer has submitted a commitment.
        let proposer = self
            .committee
            .transaction_scheduler(self.round)
            .map_err(|_| Error::NoCommittee)?;
        let proposer_commit = self.execute_commitments.get(&proposer.public_key);
        if proposer_commit.is_none() && did_timeout {
            return Err(Error::NoProposerCommitment.into());
        }

        match self.discrepancy {
            false => {
                // Discrepancy detection.
                let allowed_stragglers = self.runtime.executor.allowed_stragglers;

                // If it is already known that the number of valid commitments will not exceed the required
                // threshold, there is no need to wait for the timer to expire. Instead, proceed directly to
                // the discrepancy resolution mode, regardless of any additional commits.
                if failures > allowed_stragglers {
                    self.discrepancy = true;
                    return Err(Error::DiscrepancyDetected.into());
                }

                // While a timer is running, all nodes are required to answer.
                let mut required = total;

                // After the timeout has elapsed, a limited number of stragglers are allowed.
                if did_timeout {
                    required -= allowed_stragglers;
                    commits -= failures // Since failures count as stragglers.
                }

                // Check if the majority has been reached.
                if commits < required || proposer_commit.is_none() {
                    return Err(Error::StillWaiting.into());
                }
            }
            true => {
                // Discrepancy resolution.
                let required = total / 2 + 1;

                // Find the commit with the highest number of votes.
                let mut top_vote = &Vote::default();
                for v in votes.values() {
                    if v.tally > top_vote.tally {
                        top_vote = v;
                    }
                }

                // Fail the round if the majority cannot be reached due to insufficient votes remaining
                // (e.g. too many nodes have failed),
                let remaining = total - commits;
                if top_vote.tally + remaining < required {
                    return Err(Error::InsufficientVotes.into());
                }

                // Check if the majority has been reached.
                if top_vote.tally < required || proposer_commit.is_none() {
                    if did_timeout {
                        return Err(Error::InsufficientVotes.into());
                    }
                    return Err(Error::StillWaiting.into());
                }

                let proposer_commit = proposer_commit.expect("proposer commit should be set");
                let top_vote_commit = top_vote.commit.expect("top vote commit should be set");

                // Make sure that the majority commitment is the same as the proposer commitment.
                if !proposer_commit.mostly_equal(top_vote_commit) {
                    return Err(Error::BadProposerCommitment.into());
                }
            }
        }

        // We must return the proposer commitment as that one contains additional data.
        let proposer_commit = proposer_commit.expect("proposer commit should be set");
        Ok(proposer_commit)
    }
}

#[cfg(test)]
mod tests {
    use anyhow::{anyhow, Result};

    use crate::{
        common::{
            crypto::{
                hash::Hash,
                signature::{self, PublicKey, Signature},
            },
            namespace::Namespace,
            versioned::Versioned,
        },
        consensus::{
            registry::{
                ExecutorParameters, Node, NodeRuntime, Runtime, RuntimeGovernanceModel,
                RuntimeKind, TEEHardware,
            },
            roothash::{
                Block, ComputeResultsHeader, Error, ExecutorCommitment, ExecutorCommitmentFailure,
                ExecutorCommitmentHeader, HeaderType, Message, Pool, RegistryMessage,
                StakingMessage,
            },
            scheduler::{Committee, CommitteeKind, CommitteeNode, Role},
            staking::Transfer,
        },
    };

    use super::NodeLookup;

    struct StaticNodeLookup {
        runtime: NodeRuntime,
    }

    impl NodeLookup for StaticNodeLookup {
        fn node(&self, id: PublicKey) -> Result<Node, Error> {
            Ok(Node {
                id,
                runtimes: Some(vec![self.runtime.clone()]),
                ..Default::default()
            })
        }
    }

    #[test]
    fn test_pool_single_commitment() {
        let chain_context = "test: oasis-core tests".to_owned();

        // Generate a non-TEE runtime.
        let id =
            Namespace::from("0000000000000000000000000000000000000000000000000000000000000000");

        let rt = Runtime {
            id,
            kind: RuntimeKind::KindCompute,
            tee_hardware: TEEHardware::TEEHardwareInvalid,
            executor: ExecutorParameters {
                max_messages: 32,
                ..Default::default()
            },
            governance_model: RuntimeGovernanceModel::GovernanceEntity,
            ..Default::default()
        };

        // Generate a commitment signing key.
        let sk = signature::PrivateKey::generate();

        // Generate a committee.
        let committee = Committee {
            kind: CommitteeKind::ComputeExecutor,
            members: vec![CommitteeNode {
                role: Role::Worker,
                public_key: sk.public_key(),
            }],
            runtime_id: id,
            valid_for: 0,
        };

        // Create a pool.
        let mut pool = Pool::new(rt, committee, 0);

        // Generate a commitment.
        let (child_blk, _, mut ec) = generate_executor_commitment(id, pool.round);

        let nl = StaticNodeLookup {
            runtime: NodeRuntime {
                id,
                ..Default::default()
            },
        };

        // Test invalid commitments.
        let tcs: Vec<(&str, fn(&mut ExecutorCommitment), Error)> = vec![
            (
                "BlockBadRound",
                |ec: &mut ExecutorCommitment| ec.header.header.round -= 1,
                Error::NotBasedOnCorrectBlock,
            ),
            (
                "BlockBadPreviousHash",
                |ec: &mut ExecutorCommitment| {
                    ec.header.header.previous_hash = Hash::digest_bytes(b"invalid")
                },
                Error::NotBasedOnCorrectBlock,
            ),
            (
                "MissingIORootHash",
                |ec: &mut ExecutorCommitment| ec.header.header.io_root = None,
                Error::BadExecutorCommitment,
            ),
            (
                "MissingStateRootHash",
                |ec: &mut ExecutorCommitment| ec.header.header.state_root = None,
                Error::BadExecutorCommitment,
            ),
            (
                "MissingMessagesHash",
                |ec: &mut ExecutorCommitment| ec.header.header.messages_hash = None,
                Error::BadExecutorCommitment,
            ),
            (
                "MissingInMessagesHash",
                |ec: &mut ExecutorCommitment| ec.header.header.in_msgs_hash = None,
                Error::BadExecutorCommitment,
            ),
            (
                "BadFailureIndicating",
                |ec: &mut ExecutorCommitment| {
                    ec.header.failure = ExecutorCommitmentFailure::FailureUnknown
                },
                Error::BadExecutorCommitment,
            ),
        ];

        let msg_validator = |_: &_| Ok(());
        for (name, f, expected_err) in tcs {
            let (_, _, mut invalid_ec) = generate_executor_commitment(id, pool.round);
            f(&mut invalid_ec);

            invalid_ec.node_id = sk.public_key();
            let res = invalid_ec.sign(&sk, &id, &chain_context);
            assert!(res.is_ok(), "invalid_ec.sign({})", name);

            let res = pool.add_executor_commitment(
                &child_blk,
                &nl,
                invalid_ec,
                &msg_validator,
                &chain_context,
            );
            assert!(res.is_err(), "add_executor_commitment({})", name);
            assert_eq!(
                res.err().unwrap().to_string(),
                expected_err.to_string(),
                "add_executor_commitment({})",
                name
            );
        }

        // Generate a valid commitment.
        ec.node_id = sk.public_key();
        let res = ec.sign(&sk, &id, &chain_context);
        assert!(res.is_ok(), "ec.sign");

        // There should not be enough executor commitments.
        let res = pool.process_commitments(false);
        assert_eq!(
            res.err().unwrap().to_string(),
            Error::StillWaiting.to_string(),
            "process_commitments",
        );

        let res = pool.process_commitments(true);
        assert_eq!(
            res.err().unwrap().to_string(),
            Error::NoProposerCommitment.to_string(),
            "process_commitments",
        );

        // Test message validator function.
        let mut ec_with_msgs = ec.clone();
        ec_with_msgs.messages = vec![
            Message::Staking(Versioned {
                version: 0,
                inner: StakingMessage::Transfer(Transfer::default()),
            }),
            Message::Registry(Versioned {
                version: 0,
                inner: RegistryMessage::UpdateRuntime(Runtime::default()),
            }),
        ];
        let msg_hash = Message::messages_hash(&ec_with_msgs.messages);
        ec_with_msgs.header.header.messages_hash = Some(msg_hash);

        let res = ec_with_msgs.sign(&sk, &id, &chain_context);
        assert!(res.is_ok(), "ec_with_msgs.sign");

        let error_msg = "message validation error";
        let always_fail_msg_validator = |_: &_| -> Result<()> { Err(anyhow!(error_msg)) };
        let res = pool.add_executor_commitment(
            &child_blk,
            &nl,
            ec_with_msgs,
            &always_fail_msg_validator,
            &chain_context,
        );
        assert!(res.is_err(), "add_executor_commitment");
        assert_eq!(
            res.err().unwrap().to_string(),
            error_msg,
            "add_executor_commitment",
        );

        // Adding a commitment should succeed.
        let res = pool.add_executor_commitment(
            &child_blk,
            &nl,
            ec.clone(),
            &msg_validator,
            &chain_context,
        );
        assert!(res.is_ok(), "add_executor_commitment");

        // Adding a commitment twice for the same node should fail.
        let res = pool.add_executor_commitment(
            &child_blk,
            &nl,
            ec.clone(),
            &msg_validator,
            &chain_context,
        );
        assert!(res.is_err(), "add_executor_commitment, duplicate");

        // There should be enough executor commitments and no discrepancy.
        let res = pool.process_commitments(false);
        assert!(res.is_ok(), "process_commitments");
        let dd_ec = res
            .unwrap()
            .to_dd_result()
            .downcast_ref::<ExecutorCommitment>();
        assert_eq!(dd_ec, Some(&ec), "DD should return the correct commitment");
        assert_eq!(false, pool.discrepancy);
    }

    fn generate_executor_commitment(
        id: Namespace,
        round: u64,
    ) -> (Block, Block, ExecutorCommitment) {
        let child_blk = Block::new_genesis_block(id, round);
        let parent_blk = Block::new_empty_block(&child_blk, 1, HeaderType::Normal);

        // TODO: Add tests with some emitted messages.
        let msgs_hash = Message::messages_hash(&vec![]);
        // TODO: Add tests with some incoming messages.
        let in_msgs_hash = Message::in_messages_hash(&vec![]);

        let ec = ExecutorCommitment {
            header: ExecutorCommitmentHeader {
                header: ComputeResultsHeader {
                    round: parent_blk.header.round,
                    previous_hash: parent_blk.header.previous_hash,
                    io_root: Some(parent_blk.header.io_root),
                    state_root: Some(parent_blk.header.state_root),
                    messages_hash: Some(msgs_hash),
                    in_msgs_hash: Some(in_msgs_hash),
                    in_msgs_count: 0,
                },
                failure: ExecutorCommitmentFailure::FailureNone,
                rak_signature: None,
            },
            node_id: PublicKey::default(),
            signature: Signature::default(),
            messages: vec![],
        };

        (child_blk, parent_blk, ec)
    }
}
